### DISCUSSION TITLE
max=0 disables rate limiter :(

### DISCUSSION THREAD

**Discussion by pierregangloff** [nfriedly/express-rate-limit#259] [Ideas]
express-rate-limit is used in the Highcharts-export-server which I plan to use. But I'd love it if the rate limiter could be used as an authorization token. I.e. if we set max=0, then it NEVER allows a request to go through = it always requires the secret token. Right now, setting the max=0 disables the rate limiter. I could set it to 1, but then some requests would go through without requiring the authorization token. It would be great if max=0 meant no requests unless token is present. And max=-1 means disabled. Just a suggestion.

**Comment by nfriedly** [comment]
Yea, that "max 0 means disabled" is a bit of an unfortunate wart in the API. I could probably change it in a major release. 

[Edit: this was changed in the 7.0.0 release - setting `max=0` will now prevent any requests from getting through.]

However, for your use-case, there should probably be some _other_ level of authorization that detects the lack of a token and rejects the request with a different error message and status code (probably 401 or 403). Using a ratelimiter for that is bound to be confusing to an end-user that just happens to have a typo in their code or whatever and gets rejected on the first request.

**Reply by pierregangloff** [reply]
I agree with your assessment, this is probably a breaking change that might cause too much confusion. I ended up using the docker highcharts-export-server and adding a docker sidecar (nginx reverse proxy with njs) to validate the presence of our own API KEY. The docker+sidecar gets the current and previous valid API keys from env vars, and the sidecar makes sure that the API KEYs are there and valid before it does the proxypass to highcharts. Upon startup, on mac/win, I can pass the env vars (current and previous API KEYs) using -e, or, on AWS Fargate, the task passes the env vars into the docker container (first retrieving them from AWS SecretsManager). Also, I always pass in the AWSCURRENT and AWSPREVIOUS API Keys (secrets) so that when I rotate the API key, I don't need to cause an outage: the previous one is still passed in along with the new one. The only important thing to do is to always rotate the running docker images after I change the API Keys (never do TWO api key changes back to back without rotating new docker containers into place (otherwise the docker container is not aware of the old nor the new key). I should also mention that, upon rotation, the sidecar docker container recreates the API KEY validation logic js (I use Nginx's njs which allows me to do some basic logic in Nginx to validate the API KEYs). So, on restarting the docker containers, I have a docker-entrypoint.sh which recreates this validation.js file where the new keys are replaced (using the envsubst Linux command). It might feel a little complex but it's very straightforward and this Nginx reverse proxy sidecar is a common pattern in the docker world (in non-docker, people could put a regular reverse-proxy to do the same thing using apache or nginx). Lots of learning in the process :) But now the authorization code is fully outside the protected app (Highcharts) and can be reused for other apps (no matter if nodejs, springboot, scala, or whatever else I protect behind the Nginx reverse-proxy sidecar).

**Reply by pierregangloff** [reply]
if the highcharts project ever codes this, instead of a breaking change, they could totally new switch. For example: --api-key-required=[true|false] (defaults to false). This would make it backwards compatible (the api key is never required). If people want to use it, they set that to "true" and the server now makes sure that the API KEY is there and valid, otherwise they get a 401 Access Denied (UnauthorizedL API Key missing or invalid). So, yeah, agreed, this has no place in a rate limiter! Thanks again!
